core/src/main/java/com/linecorp/armeria/common/stream/AsyncMapStreamMessage.java

```diff
@@ -0,0 +1,241 @@
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.common.stream;
+
+import static java.util.Objects.requireNonNull;
+
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.function.Function;
+
+import org.reactivestreams.Subscriber;
+import org.reactivestreams.Subscription;
+
+import com.linecorp.armeria.common.annotation.Nullable;
+import com.linecorp.armeria.internal.common.stream.StreamMessageUtil;
+
+import io.netty.util.concurrent.EventExecutor;
+
+final class AsyncMapStreamMessage<T, U> implements StreamMessage<U> {
+    private final StreamMessage<T> source;
+    private final Function<T, CompletableFuture<U>> function;
+
+    @SuppressWarnings("unchecked")
+    AsyncMapStreamMessage(StreamMessage<? extends T> source,
+                          Function<? super T, ? extends CompletableFuture<? extends U>> function) {
+        requireNonNull(source, "source");
+        requireNonNull(function, "function");
+
+        this.source = (StreamMessage<T>) source;
+        this.function = (Function<T, CompletableFuture<U>>) function;
+    }
+
+    @Override
+    public boolean isOpen() {
+        return source.isOpen();
+    }
+
+    @Override
+    public boolean isEmpty() {
+        return source.isEmpty();
+    }
+
+    @Override
+    public long demand() {
+        return source.demand();
+    }
+
+    @Override
+    public CompletableFuture<Void> whenComplete() {
+        return source.whenComplete();
+    }
+
+    @Override
+    public void subscribe(Subscriber<? super U> subscriber, EventExecutor executor,
+                          SubscriptionOption... options) {
+        requireNonNull(subscriber, "subscriber");
+        requireNonNull(executor, "executor");
+        requireNonNull(options, "options");
+
+        source.subscribe(new AsyncMapSubscriber<>(subscriber, function, executor), executor, options);
+    }
+
+    @Override
+    public void abort() {
+        source.abort();
+    }
+
+    @Override
+    public void abort(Throwable cause) {
+        requireNonNull(cause, "cause");
+        source.abort(cause);
+    }
+
+    private static final class AsyncMapSubscriber<T, U> implements Subscriber<T>, Subscription {
+        private final Subscriber<? super U> downstream;
+        private final Function<T, CompletableFuture<U>> function;
+        private final EventExecutor executor;
+
+        @Nullable
+        private volatile Subscription upstream;
+        private volatile boolean canceled;
+        private volatile AtomicLong requestedByDownstream;
```

Comment:
```suggestion
        private final AtomicLong requestedByDownstream;
```

Reply:
Or I think we can probably use `long` without `volatile` and access this field using the `executor`. You might want to take a look at how `DeferredStreamMessage` is implemented. Could you try that please?